module/state_synchronization/storage.go

```diff
@@ -0,0 +1,996 @@
+			i++
+		}
+
+		numExecutionDataStatuses = i
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	latestIncorporatedHeight, err := s.GetLatestIncorporatedHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest incorporated height: %w", err)
+	}
+
+	if numExecutionDataStatuses > 0 {
+		if numBlobTreeRecords == 0 {
+			// there should be at least one blob tree record per tracked execution data
+			multierror.Append(errors, fmt.Errorf("found %d execution data statuses but no blob tree records", numExecutionDataStatuses))
+		} else if maxExecutionDataStatusHeight != highestBlobTreeRecordHeight {
+			multierror.Append(errors, fmt.Errorf(
+				"max execution data status height %d did not match highest blob tree record height %d",
+				maxExecutionDataStatusHeight,
+				highestBlobTreeRecordHeight,
+			))
+		}
+
+		if minExecutionDataStatusHeight <= latestIncorporatedHeight {
+			// once an execution data is incorporated, the execution data status entry should be removed
+			multierror.Append(errors, fmt.Errorf(
+				"min execution data status height %d is less than or equal to latest incorporated height %d",
+				minExecutionDataStatusHeight,
+				latestIncorporatedHeight,
+			))
+		} else if minExecutionDataStatusHeight == latestIncorporatedHeight+1 && firstExecutionDataStatus == executionDataStatusComplete {
+			// the latest incorporated height value is lower than it should be
+			multierror.Append(errors, fmt.Errorf("latest incorporated height is %d, but execution data status at height %d is complete", latestIncorporatedHeight, minExecutionDataStatusHeight))
+		}
+	} else if numBlobTreeRecords > 0 && latestIncorporatedHeight != highestBlobTreeRecordHeight {
+		multierror.Append(errors, fmt.Errorf(
+			"latest incorporated height %d did not match highest blob tree record height %d",
+			latestIncorporatedHeight,
+			highestBlobTreeRecordHeight,
+		))
+	}
+
+	return errors, nil
+}
+
+func (s *Storage) checkBlobs(blobCids map[cid.Cid]uint64) (*multierror.Error, error) {
+	blobPrefix := network.BlobServiceDatastoreKeyPrefix
+	var errors *multierror.Error
+
+	if err := s.db.View(func(txn *badger.Txn) error {
+		it := txn.NewIterator(badger.IteratorOptions{
+			PrefetchValues: false,
+			Prefix:         blobPrefix,
+		})
+		defer it.Close()
+
+		for it.Seek(blobPrefix); it.ValidForPrefix(blobPrefix); it.Next() {
+			item := it.Item()
+
+			c, err := network.CidFromBlobServiceDatastoreKey(datastore.NewKey(string(item.Key())))
+			if err != nil {
+				return fmt.Errorf("malformed blob service datastore key %v: %w", item.Key(), err)
+			}
+
+			_, ok := blobCids[c]
+			if !ok {
+				multierror.Append(errors, fmt.Errorf("no corresponding blob tree record for blob %s", c.String()))
+			}
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return errors, nil
+}
+
+// Check checks the integrity of the storage and returns any inconsistencies that were found.
+func (s *Storage) Check(options CheckOptions) error {
```

Comment:
Can you explain why we need to check the storage? 

I understand it's for integrity. But I wonder what could cause the inconsistency. 

The database (the storage layer) is supposed to guarantee data integrity as long as its public method is used to store data. 

Can you provide an example that could possibly cause inconsistency?

We could simply a lot here if the storage could provide consistency guarantee.

Reply:
So in theory, if I've implemented things correctly, there should be no inconsistency :)

But I don't think it's a smart idea to assert this just by analyzing the code, and without actually doing any tests.

This function serves to actually validate that the consistency is maintained.

Reply:
It is used in all the tests, including the ones which I haven't pushed yet. I think there is no harm making it part of the 
API as well. Nodes can run this check once upon each restart, just as a sanity check.

Reply:
It's not clear to me what needs to be stored in storage. The ExecutionData is stored in BitSwap. To me, we only need to store the height for the highest incorporated execution data, which is just a height. 

So can you explain what needs to be stored in storage, this helps me better understand why they would get inconsistent and why we need to check.

Reply:
Bitswap is not a data store, it is just a data transfer protocol. The actual data is stored in a Blockstore, which in our case is backed by badger. Bitswap requires you to pass in a [Blockstore](https://pkg.go.dev/github.com/ipfs/go-ipfs-blockstore) implementation which it uses to check if we have data requested by peers, and to store data that we've requested once we receive it. The implementation we use is https://pkg.go.dev/github.com/ipfs/go-ds-badger.

The key thing to understand here is that the underlying badger db used by the Blockstore implementation that we pass to Bitswap, is the same db that we use in `Storage`. This is so that we can prune (remove) data blobs while simultaneously updating our tracking metadata (blob records, blob tree records, etc) in a transactionally consistent manner. 

For what it's worth, I am very open to suggestions on how we can decouple the two DB's while still guaranteeing consistency. So far I haven't been able to think of anything.